llama_cpp/llama_chat_format.py

```diff
@@ -1686,9 +1687,731 @@ def message_to_str(msg: llama_types.ChatCompletionRequestMessage):
+                    calls.append(dict(name=function_name, arguments=function_args))
+
+        assert "usage" in completion
+
+        tool_calls: List[llama_types.ChatCompletionMessageToolCall] = []
+        for curr_call in calls:
+            tool_calls.append(
+                {
+                    "id": "call_"
+                    + "".join(
+                        [
+                            random.choice(string.ascii_letters + string.digits)
+                            for _ in range(24)
+                        ]
+                    ),
+                    "type": "function",
+                    "function": {
+                        "name": curr_call["name"], "arguments": curr_call["arguments"]
+                    },
+                }
+            )
+
+        function_call_dict: Union[Dict[str, str], Dict[Literal["function_call"], llama_types.ChatCompletionRequestAssistantMessageFunctionCall]] = {}
+        if len(tool_calls) > 0:
+            if self.tools is not None:
+                function_call_dict["tool_calls"] = tool_calls
+            else:
+                function_call_dict["function_call"] = {
+                    "name": tool_calls[0]["function"]["name"],
+                    "arguments": tool_calls[0]["function"]["arguments"],
+                }
+        completion["usage"]["completion_tokens"] = completion_tokens
+        return llama_types.CreateChatCompletionResponse(
+            id="chat" + completion["id"],
+            object="chat.completion",
+            created=completion["created"],
+            model=completion["model"],
+            choices=[
+                {
+                    "index": 0,
+                    "logprobs": completion["choices"][0]["logprobs"],
+                    "message": {
+                        "role": "assistant",
+                        "content": None if content == "" else content,
+                        **function_call_dict,
+                    },
+                    "finish_reason": "tool_calls" if len(tool_calls) > 0 else "stop",
+                }
+            ],
+            usage=completion["usage"],
+        )
+    
+    def generate_streaming(self, prompt: str, tool_func_choice: Optional[Union[str, Dict]]):
+        
+        def yield_response(id, created, model, finish_reason, logprobs, delta):
+            return llama_types.CreateChatCompletionStreamResponse(
+                id="chat" + id,
+                object="chat.completion.chunk",
+                created=created,
+                model=model,
+                choices=[
+                    {
+                        "index": 0,
+                        "finish_reason": finish_reason,
+                        "logprobs": logprobs,
+                        "delta": delta
+                    }
+                ],
+            )
+        
+        chunk_id, chunk_created = None, None
```

Comment:
Given these are only used in the `If "auto"` case, push the scope of the variables inside that if statement.